org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java

```diff
@@ -0,0 +1,185 @@
+			SharedSettings settings, List<CodeAction> codeActions) throws BadLocationException {
+
+		DOMDocumentType docType = document.getDoctype();
+		Position docTypeEnd = document.positionAt(docType.getEnd());
+		Position insertPosition = getEntityInsertPosition(document);
+		String message = "Declare ENTITY " + entityName;
+		String delimiter = document.lineDelimiter(insertPosition.getLine());
+
+		StringBuilder insertString = new StringBuilder();
+		if (docType.getInternalSubset() == null) {
+			String doctypeContent = docType.getTextContent();
+			// Add subset
+			if (Character.isWhitespace(doctypeContent.charAt(doctypeContent.length() - 2))) {
+				insertString.append("[");
+			} else {
+				insertString.append(" [");
+			}
+		}
+
+		if (insertPosition.getCharacter() != 0) {
+			insertString.append(delimiter);
+		}
+
+		insertString.append(createEntityDeclaration(entityName, settings));
+
+		if (docType.getInternalSubset() == null) {
+			insertString.append(delimiter + "]");
+		} else if (docTypeEnd.getLine() == insertPosition.getLine()) {
+			insertString.append(delimiter);
+		}
+
+		CodeAction action = CodeActionFactory.insert(message, insertPosition, insertString.toString(),
+				document.getTextDocument(), diagnostic);
+		codeActions.add(action);
+	}
+
+	private Position getEntityInsertPosition(DOMDocument document) throws BadLocationException {
+		TextDocument textDocument = document.getTextDocument();
+		DOMDocumentType docType = document.getDoctype();
+
+		String subset = docType.getInternalSubset();
+		if (subset == null) { // subset ([]) does not exist
+			return textDocument.positionAt(docType.getEnd() - 1);
+		}
+
+		DOMNode lastChild = docType.getLastChild();
+		if (lastChild != null) {
+			return textDocument.positionAt(lastChild.getEnd());
+		}
+
+		// empty subset exists
+		String subsetValue = "[" + subset + "]";
+		int index = docType.getTextContent().indexOf(subsetValue);
+		if (index >= 0) {
+			// if statement should always satisfy
+			// index of ]
+			index += subsetValue.length() - 1;
+			return docType.getOwnerDocument().positionAt(index + docType.getStart());
+		}
+
+		return null;
+	}
+
+	private void addDoctypeAndEntityCodeAction(String entityName, Diagnostic diagnostic, DOMDocument document,
+			SharedSettings settings, List<CodeAction> codeActions) throws BadLocationException {
+		Position insertPosition;
+		if (document.hasProlog()) {
+			insertPosition = new Position(1, 0);
+		} else {
+			insertPosition = new Position(0, 0);
+		}
+
+		String delimiter = document.lineDelimiter(insertPosition.getLine());
+		String message = "Declare DOCTYPE containing ENTITY " + entityName;
+		StringBuilder insertString = new StringBuilder();
+		insertString.append("<!DOCTYPE article [" + delimiter);
+		insertString.append(createEntityDeclaration(entityName, settings));
+		insertString.append(delimiter);
+		insertString.append("]>" + delimiter);
```

Comment:
TBH is it okay to use string concatenation instead of StringBuilder in this PR? I think StringBuilder might not be necessary unless I am appending string content in a loop.

https://stackoverflow.com/questions/4645020/when-to-use-stringbuilder-in-java
https://stackoverflow.com/questions/5234147/why-stringbuilder-when-there-is-string
https://stackoverflow.com/questions/24571463/when-should-you-explicitly-use-a-stringbuilder

Reply:
To generate XML, it should be better to use XMLBuilder instead of StringBuilder.